File: src/SemanticTypeForgers/forger.go
package forger

import (
	"reflect"
	"regexp"
	"strconv"
	"strings"
)

type AtomicBehavior[T any] struct {
	Name string
}

func NewAtomicBehavior[T any](name string) *AtomicBehavior[T] {
	return &AtomicBehavior[T]{Name: name}
}

func (a *AtomicBehavior[T]) SetSemanticType(v T) T {
	return v
}

func (a *AtomicBehavior[T]) GetPrimitiveType(v T) T {
	return v
}

func (a *AtomicBehavior[T]) Validate(value any) bool {
	return false
}

func (a *AtomicBehavior[T]) primitiveStringToValue(val string) any {
	trimmed := strings.TrimSpace(val)
	lower := strings.ToLower(trimmed)

	if trimmed == "" {
		return val
	}
	if val == "{}" || val == "[]" {
		return 0
	}
	if lower == "true" {
		return true
	}
	if lower == "false" {
		return false
	}
	if lower == "null" || lower == "undefined" {
		return nil
	}

	re := regexp.MustCompile(`(?i)^[+-]?(?:(?:\d+\.?\d*)|(?:\.\d+))(?:e[+-]?\d+)?$`)
	if re.MatchString(trimmed) {
		if strings.Contains(trimmed, ".") || strings.Contains(lower, "e") {
			if f, err := strconv.ParseFloat(trimmed, 64); err == nil {
				return f
			}
		} else {
			if i, err := strconv.ParseInt(trimmed, 10, 64); err == nil {
				return int(i)
			}
		}
	}

	return val
}

func (a *AtomicBehavior[T]) unwrap(val any, seen map[uintptr]bool, depth int) any {
	if depth > 100 {
		return nil
	}
	if val == nil {
		return nil
	}

	switch v := val.(type) {
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64, bool:
		return v
	case string:
		return a.primitiveStringToValue(v)
	}

	rt := reflect.TypeOf(val)
	rv := reflect.ValueOf(val)

	if rt.Kind() == reflect.Ptr {
		ptr := rv.Pointer()
		if seen[ptr] {
			return nil
		}
		seen[ptr] = true
		if !rv.IsNil() {
			return a.unwrap(rv.Elem().Interface(), seen, depth+1)
		}
		return nil
	}

	if rt.Kind() == reflect.Slice || rt.Kind() == reflect.Array {
		if rv.Len() == 0 {
			return nil
		}
		return a.unwrap(rv.Index(0).Interface(), seen, depth+1)
	}

	if rt.Kind() == reflect.Map {
		if rv.Len() == 0 {
			return 0
		}
		keys := rv.MapKeys()
		return a.unwrap(rv.MapIndex(keys[0]).Interface(), seen, depth+1)
	}

	return nil
}

func (a *AtomicBehavior[T]) ConvertToPrimitive(value any) any {
	return a.unwrap(value, make(map[uintptr]bool), 0)
}

func (a *AtomicBehavior[T]) Forge(v T) T {
	return v
}

func (a *AtomicBehavior[T]) ProcessValue(value any) any { // Use 'any' return type for flexibility in testing
	if a.Validate(value) {
		return value
	}

	if m, ok := value.(map[string]any); ok {
		var pPrice, pDiscount, dPrice, pFees float64
		hasPrice := false

		if val, ok := m["productPrice"]; ok {
			prim := a.ConvertToPrimitive(val)
			if f, ok := prim.(float64); ok {
				pPrice = f
				hasPrice = true
			} else if i, ok := prim.(int); ok {
				pPrice = float64(i)
				hasPrice = true
			}
		}

		if val, ok := m["productDiscount"]; ok {
			prim := a.ConvertToPrimitive(val)
			if f, ok := prim.(float64); ok {
				pDiscount = f
			} else if i, ok := prim.(int); ok {
				pDiscount = float64(i)
			}
		}

		if val, ok := m["deliveryPrice"]; ok {
			prim := a.ConvertToPrimitive(val)
			if f, ok := prim.(float64); ok {
				dPrice = f
			} else if i, ok := prim.(int); ok {
				dPrice = float64(i)
			}
		}

		if val, ok := m["paymentFees"]; ok {
			prim := a.ConvertToPrimitive(val)
			if f, ok := prim.(float64); ok {
				pFees = f
			} else if i, ok := prim.(int); ok {
				pFees = float64(i)
			}
		}

		if hasPrice {
			finalPrice := pPrice - pDiscount + dPrice + pFees
			return finalPrice
		}
	}

	return a.ConvertToPrimitive(value)
}

File: src/SemanticTypeForgers/forger.pl
:- module(forger, [get_primitive_type/2, validate/2, convert_to_primitive/2, process_value/2]).

get_primitive_type(V, V).

validate(_, false).

primitive_string_to_value(Val, Res) :-
    normalize_space(atom(Trimmed), Val),
    downcase_atom(Trimmed, Lower),
    ( Trimmed == '' -> Res = Val
    ; (Val == '{}' ; Val == '[]') -> Res = 0
    ; Lower == 'true' -> Res = true
    ; Lower == 'false' -> Res = false
    ; (Lower == 'null' ; Lower == 'undefined') -> Res = null
    ; catch(atom_number(Trimmed, Num), _, fail) -> Res = Num
    ; Res = Val
    ).

convert_to_primitive(Val, Res) :-
    ( atom(Val), \+ number(Val) -> primitive_string_to_value(Val, Res)
    ; string(Val) -> atom_string(Atom, Val), primitive_string_to_value(Atom, Res)
    ; Res = Val
    ).

forge(V, V).

process_value(Dict, Res) :-
    is_dict(Dict),
    get_dict(productPrice, Dict, PPriceRaw),
    convert_to_primitive(PPriceRaw, PPrice),
    ( get_dict(productDiscount, Dict, PDiscountRaw) -> convert_to_primitive(PDiscountRaw, PDiscount) ; PDiscount = 0 ),
    ( get_dict(deliveryPrice, Dict, DPriceRaw) -> convert_to_primitive(DPriceRaw, DPrice) ; DPrice = 0 ),
    ( get_dict(paymentFees, Dict, PFeesRaw) -> convert_to_primitive(PFeesRaw, PFees) ; PFees = 0 ),
    number(PPrice),
    FinalPrice is PPrice - PDiscount + DPrice + PFees,
    forge(FinalPrice, Res).
process_value(Val, Res) :-
    \+ is_dict(Val),
    convert_to_primitive(Val, Prim),
    forge(Prim, Res).
